src/db.c

```diff
@@ -48,6 +48,46 @@
 int expireIfNeeded(redisDb *db, robj *key, int flags);
 int keyIsExpired(redisDb *db, robj *key);
 
+/* Returns next dictionary from the iterator, or NULL if iteration is complete. */
+dict *dbNextDict(dbIterator *dbit) {
+    while (dbit->index < dbit->db->dict_count - 1) {
+        dbit->index++;
+        dict *d = dbit->db->dict[dbit->index];
+        /* There is a single dictionary in non cluster mode,
+         * in cluster mode return first non-empty sub-dictionary. */
+        if (!server.cluster_enabled || dictSize(d) > 0) return d;
+    }
+    return NULL;
+}
+
+void dbInitIteratorAt(dbIterator *dbit, redisDb *db, int slot) {
+    serverAssert(slot == 0 || server.cluster_enabled);
+    dbit->db = db;
+    dbit->index = slot - 1; /* Start one slot ahead, as dbNextDict increments index right away. */
```

Comment:
How about deferring the increment of dbit->index, maybe one day we need to traverse backwards.
```suggestion
    while (dbit->index < dbit->db->dict_count) {
        dict *d = dbit->db->dict[dbit->index++];
        /* There is a single dictionary in non cluster mode,
         * in cluster mode return first non-empty sub-dictionary. */
        if (!server.cluster_enabled || dictSize(d) > 0) return d;
    }
    return NULL;
}

void dbInitIteratorAt(dbIterator *dbit, redisDb *db, int slot) {
    serverAssert(slot == 0 || server.cluster_enabled);
    dbit->db = db;
    dbit->index = slot; /* Start one slot ahead, as dbNextDict increments index right away. */
```

Reply:
This function has changed since the comment was posted, I think it should no longer be a concern.